Add multivariate_normal and pdf to jax.scipy.stats

h3jia opened the pull request:
1. Add `pdf` of the existing distributions in `jax.scipy.stats`
2. Add `logpdf` and `pdf` of `multivariate_normal`

h3jia committed:
add scipy.stats.norm.pdf

h3jia committed:
add scipy.stats.beta.pdf

h3jia committed:
add scipy.stats.expon.pdf

h3jia committed:
add scipy.stats.gamma.pdf

h3jia committed:
add scipy.stats.laplace.pdf

h3jia committed:
add scipy.stats.uniform.pdf

h3jia committed:
Update __init__.py

h3jia committed:
add multivariate_normal logpdf and pdf

_promote_args_like in the first line of logpdf does not work well, so I temporarily comment it out.

mattjj commented:
Thanks for your contribution!

Could you add tests for these new functions, similar to the existing ones in tests/scipy_stats_test.py or tests/lax_scipy_test.py (the former is slightly un-conventional, while the latter conforms to our typical test patterns)? Any tests, even hand-written ones, would be preferable to no tests, though if you follow our parameterized-test-cases pattern you can cover a lot of shapes and dtypes.

Let me know if you have questions or concerns about adding tests!

HerculesJack commented:
> Thanks for your contribution!
> 
> Could you add tests for these new functions, similar to the existing ones in tests/scipy_stats_test.py or tests/lax_scipy_test.py (the former is slightly un-conventional, while the latter conforms to our typical test patterns)? Any tests, even hand-written ones, would be preferable to no tests, though if you follow our parameterized-test-cases pattern you can cover a lot of shapes and dtypes.
> 
> Let me know if you have questions or concerns about adding tests!

Sure, I'll do it this weekend. : )

h3jia committed:
Update scipy_stats_test.py

1. Add testMultivariateNormalLogPdf, though currently it only tests dim=4 case
2. Add testGammaLogPdf
2. Remove some onp.abs because rand_positive is used

h3jia committed:
Fix dtype problem for x64

h3jia committed:
Update scipy_stats_test.py

h3jia committed:
python2 does not have @, so use np.dot instead

HerculesJack commented:
Sorry for my late update! 

I've added a simple test for `multivariate_normal.logpdf` and `gamma.logpdf`. Though I should mention that the test for `multivariate_normal.logpdf` is just a preliminary version, since it only tests dim=4 case. I'm not sure how to test arbitrary shapes for a multivariate distribution in a way similar to the existing ones, which are for univariate distributions.

mattjj commented on jax/scipy/stats/multivariate_normal.py:
```diff
@@ -0,0 +1,49 @@
+
+@_wraps(osp_stats.multivariate_normal.logpdf)
+def logpdf(x, mean, cov):
+  # x, mean, cov = _promote_args_like(osp_stats.multivariate_normal.logpdf, x, mean, cov)
```
Is this line not needed anymore? If not, maybe just delete it instead of leaving it commented out.

But it seems more likely that we want promotion logic here, rather than casting everything to cov.dtype. Can we just put this line back in (replacing the `astype` lines immediately below), or is there a problem with it?

HerculesJack replied:
I commented that line out because it led to the following error. I think it's because the positive-definite check in `scipy.stats.multivariate_normal` brings some problem to `_promote_args_like`. I didn't know how to fix this so I used `astype` instead. Could you give me some ideas on this?

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-437963fd1205> in <module>()
----> 1 scipy.stats.multivariate_normal.pdf(np.array([1,2]),np.array([3,4]),np.eye(2))

/global/u2/h/hejia/jax/jax/scipy/stats/multivariate_normal.py in pdf(x, mean, cov)
     47 @_wraps(osp_stats.multivariate_normal.pdf)
     48 def pdf(x, mean, cov):
---> 49   return lax.exp(logpdf(x, mean, cov))

/global/u2/h/hejia/jax/jax/scipy/stats/multivariate_normal.py in logpdf(x, mean, cov)
     28 @_wraps(osp_stats.multivariate_normal.logpdf)
     29 def logpdf(x, mean, cov):
---> 30   x, mean, cov = _promote_args_like(osp_stats.multivariate_normal.logpdf, x, mean, cov)
     31   x = x.astype(cov.dtype)
     32   mean = mean.astype(cov.dtype)

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _promote_args_like(op, *args)
    195   """Convenience function to apply shape and dtype promotion to result type."""
    196   _check_arraylike(op.__name__, *args)
--> 197   return _promote_shapes(*_promote_to_result_dtype(op, *args))
    198 
    199 

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _promote_to_result_dtype(op, *args)
    167 def _promote_to_result_dtype(op, *args):
    168   """Convenience function to promote args directly to the op's result dtype."""
--> 169   to_dtype = _result_dtype(op, *args)
    170   return [lax.convert_element_type(arg, to_dtype) for arg in args]
    171 

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _result_dtype(op, *args)
    174   """Compute result dtype of applying op to arguments with given dtypes."""
    175   args = (onp.ones((0,) * ndim(arg), _dtype(arg)) for arg in args)
--> 176   return _dtype(op(*args))
    177 
    178 

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/stats/_multivariate.py in logpdf(self, x, mean, cov, allow_singular)
    485         dim, mean, cov = self._process_parameters(None, mean, cov)
    486         x = self._process_quantiles(x, dim)
--> 487         psd = _PSD(cov, allow_singular=allow_singular)
    488         out = self._logpdf(x, mean, psd.U, psd.log_pdet, psd.rank)
    489         return _squeeze_output(out)

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/stats/_multivariate.py in __init__(self, M, cond, rcond, lower, check_finite, allow_singular)
    150         # Note that eigh takes care of array conversion, chkfinite,
    151         # and assertion that the matrix is square.
--> 152         s, u = scipy.linalg.eigh(M, lower=lower, check_finite=check_finite)
    153 
    154         eps = _eigvalsh_to_eps(s, cond, rcond)

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/linalg/decomp.py in eigh(a, b, lower, eigvals_only, overwrite_a, overwrite_b, turbo, eigvals, type, check_finite)
    430         if eigvals is None:
    431             w, v, info = evr(a1, uplo=uplo, jobz=_job, range="A", il=1,
--> 432                              iu=a1.shape[0], overwrite_a=overwrite_a)
    433         else:
    434             (lo, hi) = eigvals

ValueError: failed to create intent(cache|hide)|optional array-- must have defined dimensions but got (0,)
```

mattjj replied:
Sorry for being slow to respond on this one.

I see, thanks for explaining. Could you add a short comment to the code with this information? Maybe something like:

```
# TODO add back dtype promotion after resolving the PSD-check issue
# see https://github.com/google/jax/pull/268
```

mattjj replied:
On second thought, I can add that TODO note myself after we merge. No need to slow this down :)

mattjj commented on jax/scipy/stats/multivariate_normal.py:
```diff
@@ -0,0 +1,49 @@
+      subtract(x_2d, mean)).reshape(x_shape).astype(cov.dtype)
+  else:
+    quadratic = dot(dot(subtract(x, mean), inv(cov)), subtract(x, mean).T).astype(cov.dtype)
+  return lax.div(lax.neg(lax.add(log_normalizer, quadratic)), two)
```
If it's more convenient, you could use the infix operators here (which forward to numpy functions) rather than calling `lax` functions directly, e.g.

```python
return -(log_normalizer + quadratic) / two
```

The only reason to prefer the `lax` functions over the numpy-based infix operators sometimes is that they have more stringent dtype-casting and shape-broadcasting rules, which can catch errors and make the Python tracing execute a miniscule amount faster.

To be clear, it's okay to keep this how it is. I just wanted to emphasize that you can use whatever seems most convenient.

mattjj reviewed (changes requested):
Excellent, thank you for this contribution!

I marked this as "request changes" just for a single commented-out line in multivariate_normal.py (the one that would call `_promote_args_like`). Once we get that finalized, this is good to merge!

mattjj reviewed (approved):
Looks good to me! Thanks for this contribution.
